src/Mod/Sketcher/Gui/DrawSketchHandler.cpp

```diff
@@ -420,137 +420,130 @@ DrawSketchHandler::suggestedConstraintsPixmaps(std::vector<AutoConstraint>& sugg
-                                          AutoConstraint::TargetType type)
+DrawSketchHandler::PreselectionData DrawSketchHandler::getPreselectionData()
 {
-    suggestedConstraints.clear();
     SketchObject* obj = sketchgui->getSketchObject();
 
-    if (!sketchgui->Autoconstraints.getValue()) {
-        return 0;  // If Autoconstraints property is not set quit
-    }
-    // direction of hit shape (if it is a line, the direction of the line)
-    Base::Vector3d hitShapeDir = Base::Vector3d(0, 0, 0);
-    bool preselectIsLine = false;
-
-    // Get Preselection
+    // Extract preselection information (vertex, curve, cross)
+    PreselectionData preSelData;
     int preSelPnt = getPreselectPoint();
     int preSelCrv = getPreselectCurve();
     int preSelCrs = getPreselectCross();
 
-    int GeoId = GeoEnum::GeoUndef;
-    PointPos PosId = PointPos::none;
-
     if (preSelPnt != -1) {
-        obj->getGeoVertexIndex(preSelPnt, GeoId, PosId);
+        obj->getGeoVertexIndex(preSelPnt, preSelData.geoId, preSelData.posId);
     }
     else if (preSelCrv != -1) {
         const Part::Geometry* geom = obj->getGeometry(preSelCrv);
-
-        // ensure geom exists in case object was called before preselection is updated
         if (geom) {
-            GeoId = preSelCrv;
+            preSelData.geoId = preSelCrv;
             if (geom->is<Part::GeomLineSegment>()) {
                 auto* line = static_cast<const Part::GeomLineSegment*>(geom);
-                hitShapeDir = line->getEndPoint() - line->getStartPoint();
-                preselectIsLine = true;
+                preSelData.hitShapeDir = line->getEndPoint() - line->getStartPoint();
+                preSelData.isLine = true;
             }
         }
     }
     else if (preSelCrs == 0) {
-        // root point
-        GeoId = Sketcher::GeoEnum::RtPnt;
-        PosId = PointPos::start;
+        preSelData.geoId = Sketcher::GeoEnum::RtPnt;
+        preSelData.posId = PointPos::start;
     }
     else if (preSelCrs == 1) {
-        // x axis
-        GeoId = Sketcher::GeoEnum::HAxis;
-        hitShapeDir = Base::Vector3d(1, 0, 0);
-        preselectIsLine = true;
+        preSelData.geoId = Sketcher::GeoEnum::HAxis;
+        preSelData.hitShapeDir = Base::Vector3d(1, 0, 0);
+        preSelData.isLine = true;
     }
     else if (preSelCrs == 2) {
-        // y axis
-        GeoId = Sketcher::GeoEnum::VAxis;
-        hitShapeDir = Base::Vector3d(0, 1, 0);
-        preselectIsLine = true;
+        preSelData.geoId = Sketcher::GeoEnum::VAxis;
+        preSelData.hitShapeDir = Base::Vector3d(0, 1, 0);
+        preSelData.isLine = true;
+    }
+    return preSelData;
+}
+
+bool DrawSketchHandler::isLineCenterAutoConstraint(int GeoId, const Base::Vector2d& Pos)
+{
+    SketchObject* obj = sketchgui->getSketchObject();
+
+    auto* geo = obj->getGeometry(GeoId);
+    if (auto* line = dynamic_cast<const Part::GeomLineSegment*>(geo)) {

```

Comment:
`isDerivedFrom<>()`?

Reply:
We actually need line after. I used this syntax (making the var in the if) because it was suggested to me before.

Reply:
I see what you mean. We previously had var assignment in if-statement, though to my knowledge, it was decided against it - though this is mostly a cosmetic thing.

Further, dynamic_cast is slow, because it type checks by looking at the vtable. isDerivedFrom checks a type property of the class, and is thus fast.

So I think this was a incorrect advice (not your fault). Unfortunately.

Reply:
Ok then I reverted this